transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -1,26 +1,21 @@
 """A background thread for asynchronous operations on the jlink adapter."""
 
+import asyncio
```

Comment:
### Overall Goal

There are two parts of the implementation needed here:

1. Fundamental interaction with `jlink` needs to be wrapped in `coroutines` for consumption by a `StandardDeviceAdapter` subclass whose API is natively coroutine based.  That is the refactor needed in `JLinkControlThread`, which should be renamed `AsyncJLink`.  All of the exposed methods on `AsyncJLink` should be coroutines.
2.  `JLinkDeviceAdapter` needs to restructured so that it subclasses `StandardDeviceAdapter` and implements the `AbstractDeviceAdapter` coroutine API.  It will delegate all of the actual work (connect to jlink, take a jlink action, etc) to `AsyncJLink` to perform and since those methods will be coroutines, the coroutines in StandardDeviceAdapter can nicely `await` them.

This draft PR is targeted at goal 1 above solely so I'll focus comments on that.

### Comments

- we have a common wrapper on top of `asyncio` that all of CoreTools should use: [`BackgroundEventLoop`](https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/utilities/async_tools/event_loop.py#L252).  

- No `asyncio` loops should be created independently in any CoreTools subsysterm.  There should be a single BackgroundEventLoop per process that is declared [here](https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/utilities/async_tools/event_loop.py#L683) (`SharedLoop`), i.e. normally asyncio has a single foreground loop per process, we have a single background loop per process.

- Any long-running blocking style interaction with the jlink hardware (pretty much any call on the `pylink` object) needs to be [run on an executor thread](https://github.com/iotile/coretools/blob/master/iotilecore/iotile/core/utilities/async_tools/event_loop.py#L559) so it doesn't block the single shared background event loop.

    You will find many such convenience wrappers around `asyncio` routines that you may choose to use on the `BackgroundEventLoop` class.

- You should not need to manually create any threads in this implementation.  The goal is to move from explicit multithreading to running all background work inside of the `asyncio` loop so we don't achieve that if we are forced to have a mix of threading / asyncio code.

### Suggestions

- In other `asyncio` ports we have found the following structure useful:

   1. Implement the class with an `async` start() and `stop()` coroutine methods.  Defer any blocking work from `__init__` to `start()` so that it can be `awaited`.  This is important since you can't declare `__init__` `async`.  For example, `start()` could connect to `pylink` and setup any necessary structures.  `stop()` could cleanly disconnect from the pylink.

   2. If you need to perform background work that no one is explicitly awaiting, such as continually polling the jlink for events, you can do that by creating a `BackgroundTask` and adding it to the `BackgroundEventLoop`.  This makes sure that it is properly cleaned up when the process exits.  You should create it in `start()` and stop it in `stop()`.  If you have a hierarchy of such tasks, `BackgroundTask` can have children but you shouldn't need that here I don't think.

   3. Always take a `loop=SharedLoop` keyword argument (with default)in `__init__` for any object that interacts with a `BackgroundEventLoop`.  This allows passing a different loop for unit testing.

   4. You should not need to often call `BackgroundEventLoop.get_loop()` to get the raw `asyncio` loop.  Generally there is a better way to achieve your goal using high level `BackgroundEventLoop` methods like `launch_coroutine` or things like `create_event`, `create_queue`, etc.

   5. Never try to have two different asnycio loops running and communicate between them.  It never works well.

Reply:
Thank you for thorough explanation

Reply:
Hi Tim! 
I wanted to ask you about the point:
_Any long-running blocking style interaction with the jlink hardware (pretty much any call on the pylink object) needs to be run on an executor thread so it doesn't block the single shared background event loop._
There are some functions, **async def send_rpc** for example, in AsyncJlink that contains a lot of jlink IO calls. Is this mandatory for AsyncJlink to provide API using coroutines only or it would be okay if **def send_rpc** will be called in ** run_in_executor ** as a whole?

Reply:
I don't think it's a hard requirement what the interface is between `AsyncJLink` and `JLinkDeviceAdapter`, though in this case I think it would actually be easier to make `send_rpc` async for the following reasons:

- `send_rpc` now is really just a `write_memory` followed by a `trigger_interrupt` and then a polling loop.  `write_memory` will already need to be `async` to call from `async debug`.  `trigger_interrupt` is easy to make async, so then it comes down to the polling loop.

   That loop is going to need to go away most likely anyway since the jlink needs to constantly be checking for new tracing or streaming data out of the device and those checks can be merged together with the checks for whether an RPC is finished into a single background asyncio task.

   So, given the need to support tracing and streaming, `send_rpc` should become a simple function whose core building block `write_memory` already got an asyncio wrapper for use elsewhere.

That being said, there is no hard requirement on what the `AsyncJLink` API is beyond that:
- there are no background threads spun up for anything (use executor calls instead if needed)
- the main asyncio loop (`SharedLoop`) is never blocked
- it works correctly

Reply:
I am just trying to understand if I am wrong. 
_send_rpc_ is just a bunch of _jlink.read_mem_ and _jlink.write_mem_ functions. These functions are considered blocking. To me it looks like just calling _send_rpc_ and other help functions in executor would resolve possible problems of blocking BackgroundLoop as well as it shows clearly intention (interaction with jlink should be async and in separate thread).
Do I miss something?   

I updated PR so I could see what I mean, I will appreciate any comments from your side.

Reply:
I understand the idea and, of course it would work, I just think that you would have a lot of places in your code with a `loop.run_in_executor(self.read_mem, ...)` whenever you wanted to read_mem.  

If you instead did something like (rename `read_mem` to `_read_mem_blocking` and then did:
```
async def read_mem():
    loop.run_in_executor(self._read_mem_blocking, ...)
```

It would be clear that `read_mem` is an async function (it's marked as such) whereas the blocking variety is also clearly marked (and internal so no one calls it by mistake.  Then you could build up an async send_rpc as just:

```
async def send_rpc():
    await async_jlink.write_mem()
    await async_jlink.trigger_interrupt()

    await async_jlink.rpc_complete_event()
```

Where `rpc_complete_event` is an `asyncio.Event()` that is being set by the background asyncio reader task that is polling for rpc completion and checking for streaming/tracing data.

Reply:
I see your point. Thank you for explanations. I updated as you suggest, still have to refactor and add handling events for rpc and streaming/tracing

Reply:
I was wondering if it's better for you to review separate PR refactoring + tracing/streaming or everything in the single PR? I made all changes regarding refactoring I wanted. Please, feel free to share you thoughts.

Reply:
Generally I would prefer smaller PRs with each feature as you did above to make it easier and quicker to review rather than a giant change set.  

I'm taking a look now at the PR and will give feedback.

Reply:
May you please create *feat-jlink-upgrade* branch. I am not sure I have access rights to do it.

Reply:
Done, branch exists now